File: Bai_99/main.py
def is_leap_year(year):
    return (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0)

def is_valid_date(day, month, year):
    if month < 1 or month > 12: return False
    days_in_month = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    if month == 2 and is_leap_year(year): days_in_month[2] = 29
    return 1 <= day <= days_in_month[month]

def process_date_string(s):
    valid_dates = []
    year = 2000 + int(s[-2:])
    if len(s) == 5:
        day_str = s[:2]
        month_str = s[2]
        day = int(day_str)
        month = int(month_str)
        if is_valid_date(day, month, year): valid_dates.append(f"{day_str}/{month_str}/{year}")
    day_str = s[0]
    month_str = s[1:len(s) - 2]
    day = int(day_str)
    month = int(month_str)
    if is_valid_date(day, month, year): valid_dates.append(f"{day_str}/{month_str}/{year}")
    return valid_dates

def main():
    with open('Ngay.inp', 'r') as input_file:
        with open('Ngay.out', 'w') as output_file:
            N = int(input_file.readline().strip())
            for _ in range(N):
                s = input_file.readline().strip()
                valid_dates = process_date_string(s)
                if not valid_dates:
                    output_file.write("NO\n")
                else:
                    output_file.write(';'.join(valid_dates) + '\n')

if __name__ == "__main__":
    main()

File: Bai_99/main.cpp
#include<bits/stdc++.h>
using namespace std;

bool isLeapYear(int year){
    return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
}

bool isValidDate(int day, int month, int year){
    if(month < 1 || month > 12) return false;
    int daysInMonth[13] = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    if(month == 2 && isLeapYear(year)) daysInMonth[2] = 29;
    return day >= 1 && day <= daysInMonth[month];
}

vector<string> processDateString(const string& s){
    vector<string> validDates;
    int year = 2000 + stoi(s.substr(s.size() - 2));
    string dayStr, monthStr; int day,month;
    if(s.size() == 5){
        dayStr = s.substr(0,2); day = stoi(dayStr);
        monthStr = s.substr(2,1); month = stoi(monthStr);
        if(isValidDate(day, month, year))
            validDates.push_back(dayStr + "/" + monthStr + "/" + to_string(year));
    }
    dayStr = s.substr(0,1); day = stoi(dayStr);
    monthStr = s.substr(1,s.size() - 3); month = stoi(monthStr);
    if(isValidDate(day, month, year))
        validDates.push_back(dayStr + "/" + monthStr + "/" + to_string(year));
    return validDates;
}

int main(){
    ifstream inputFile("Ngay.inp");
    ofstream outputFile("Ngay.out");
    int N; inputFile >> N;
    for(int i = 0; i < N; ++i) {
        string s; inputFile >> s;
        vector<string> validDates = processDateString(s);
        if(validDates.empty()) outputFile << "NO";
        for(int i = 0; i < validDates.size(); ++i){
            outputFile << validDates[i];
            if(i != validDates.size() - 1) outputFile << ';';
        }
        outputFile << endl;
    }
    inputFile.close();
    outputFile.close();
    return 0;
}
